session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift

```diff
@@ -0,0 +1,85 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import UIKit
+
+/// Single unique ID of a view in view-tree hierarchy.
+/// It is used to mark `UIViews` which correspond to single wireframe in the replay.
+internal typealias NodeID = Int64
+
+/// Pair of unique IDs of a single view in view-tree hierarchy.
+/// It is used to mark `UIViews` which correspond to two wireframes in the replay.
+internal typealias NodeID2 = (NodeID, NodeID)
+
+/// Manages `NodeIDs` for `UIView` instances.
+///
+/// Each `UIView` can be assigned one or more unique IDs. All IDs are linked to `UIView` lifespans, so querying
+/// IDs for the same instance of `UIView` will always give the same values.
+///
+/// **Note**: All `NodeIDGenerator` APIs must be called on the main thread.
+internal final class NodeIDGenerator {
+    /// Upper limit for generated IDs.
+    /// After `currentID` reaches this limit, it will start from `0`.
+    private let maxID: NodeID
+    /// Tracks next `NodeID` to assign.
+    private var currentID: NodeID
+
+    init(currentID: NodeID = 0, maxID: NodeID = .max) {
+        self.currentID = currentID
+        self.maxID = maxID
+    }
+
+    /// Returns single `NodeID` for given instance of `UIView`.
+    /// - Parameter view: the `UIView` object
+    /// - Returns: the `NodeID` of queried instance
+    func nodeID(for view: UIView) -> NodeID {
+        if let currentID = view.nodeID {
+            return currentID
+        } else {
+            let id = getNextID()
+            view.nodeID = id
+            return id
+        }
+    }
+
+    /// Returns two `NodeIDs` for given instance of `UIView`.
+    /// - Parameter view: the `UIView` object
+    /// - Returns: `NodeIDs` of queried instance
+    func nodeID2(for view: UIView) -> NodeID2 {
```

Comment:
IMO, It would be clearer if the method returns tuples explicitly.
`func nodeID2(for view: UIView) -> (NodeID, NodeID)`

Reply:
Also, this won't scale well if we have components that have other requirements.
We could have a method for requesting `n` IDs.
```func nodeIDs(for view: UIView, count: Int) -> [NodeID]```

Reply:
I'd prefer to keep it as it gives me good documentation point:
```swift
/// Single unique ID of a view in view-tree hierarchy.
/// It is used to mark `UIViews` which correspond to single wireframe in the replay.
internal typealias NodeID = Int64

/// Pair of unique IDs of a single view in view-tree hierarchy.
/// It is used to mark `UIViews` which correspond to two wireframes in the replay.
internal typealias NodeID2 = (NodeID, NodeID)
```
About scalability: it is true, but I'm not afraid of it - it's extremely unlikely that we'll need more than a few IDs per `UIView`. Even this tuple is quite experimental for now and used only to represent `UISwitch` as two wireframes (this may change if we decide to render OS-specific elements on the player side by introducing OS-specific wireframes instead of mimicking its look&feel with 2 shape WFs).

How does it sound @maxep ?

Reply:
Yes sounds good, we just need to be careful not to introduce too many `typealias`es at the module level (inner to object is fine), it can get difficult to track and be confusing.